### PR TITLE
[homuntan02] iP

### PR DESCRIPTION
**Pull request by homuntan02**
```ruby
 ____        _        
|  _ \ _   _| | _____ 
| | | | | | | |/ / _ \
| |_| | |_| |   <  __/
|____/ \__,_|_|\_\___|

```
Duke helps you keep track of things you need to do in 1 simple list. it has:

- **Straightforward** and **simple** commands
- Marker system to track which task is completed

Features include:

- [x] Managing Task
- [x] 3 different task types :
        -  To-Dos
        -  Deadlines
        -  Events

- [x] Searching for task with keywords
- [ ] Task Reminder (coming soon)

Here is the `main` method if you code in Java.
```ruby
public class Duke {
    public static void main(String[] args) {
        new Duke("./data/").run();
    }
}
```

### CODE REVIEW

File: src/main/java/duke/parsers/Parser.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+            return new DeadlineCommand(requestContent);
+
+        case "event":
+            return new EventCommand(requestContent);
+
+        case "delete":
+            if (requestContent.trim() == "") {
+                throw new MissingArgumentException("The index cannot be empty.");
+            }
+            int index = Integer.parseInt(requestContent.trim()) - 1;
```
Comment (clevon-w) [inline src/main/java/duke/parsers/Parser.java:61]: Does this handle exceptions thrown if the user passes in something like "delete something"? I think if "something" is passed into `Integer.parseInt` an error will be thrown.

File: src/main/java/duke/commands/DeadlineCommand.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+public class DeadlineCommand extends Command {
+    private String requestContent;
+    public DeadlineCommand(String requestContent) {
+        super("DEADLINE");
+        this.requestContent = requestContent;
+    }
+
+    /**
+     * Adds deadline to a task list.
+     */
```
Comment (clevon-w) [inline src/main/java/duke/commands/DeadlineCommand.java:18]: Does this require a JavaDoc? This is the only `execute` method that has a JavaDoc out of all the `Command` subclasses.

File: src/test/java/command/DeadlineCommandTest.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    }
+
+    @Test
+    public void shouldAddDeadline() {
+        DeadlineCommand demoDeadlineCommand = new DeadlineCommand("Demo /by 2002-02-12 19:00");
+        demoDeadlineCommand.execute(this.tasks, this.ui, this.storage);
+        Task actual = this.tasks.showTask(0);
+        Deadline expected = new Deadline("Demo","2002-02-12 19:00");
+        assertEquals(expected, actual);
+    }
```
Comment (clevon-w) [inline src/test/java/command/DeadlineCommandTest.java:33]: Is it a good idea to test for errors thrown? So that you can check if the right error is thrown for the right edge cases.

**Review by clevon-w — COMMENTED** [review]
Really great work Homun! Keep it up!
I just love the way you structured your repository. It made it so easy to read and understand. I have so much to learn from you!

File: src/main/java/duke/Duke.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+import duke.commands.Command;
+import duke.dukeexceptions.DukeExceptions;
+import duke.parsers.Parser;
+import duke.storage.Storage;
+import duke.tasklist.TaskList;
+
+
+/**
+ * The main duke program.
```
Comment (szejiancheng) [inline src/main/java/duke/Duke.java:14]: Perhaps the header javadoc could be more descriptive!

File: src/main/java/duke/Duke.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+            } finally {
+                ui.showLine();
+            }
+        }
+
+        storage.save(taskList);
+    }
+
+    public static void main(String[] args) {
+        new Duke("./data/").run();
```
Comment (szejiancheng) [inline src/main/java/duke/Duke.java:88]: Is there a way to make this less system specific? What happens when the filepath isnt valid?

File: src/main/java/duke/Ui.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
+package duke;
+import duke.dukeexceptions.DukeExceptions;
```
Comment (szejiancheng) [inline src/main/java/duke/Ui.java]: Consider separating these with new lines for readability

File: src/main/java/duke/commands/Command.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+public abstract class Command {
+    private boolean isEnd;
+    private String tag;
+
+    /**
+     * Constuctor for a commmand.
+     *
+     * @param tag to know what command is used.
+     */
+    public Command(String tag) {
```
Comment (szejiancheng) [inline src/main/java/duke/commands/Command.java:16]: As the tag corresponds to the subtype of Command, is there something more useful you could store in your tag?

File: src/main/java/duke/commands/EventCommand.java [hunk 8]
```diff
@@ -0,0 +1,20 @@
+package duke.commands;
+import duke.Ui;
+import duke.storage.Storage;
+import duke.tasklist.TaskList;
+
+/**
+ * Command to add an event to a task list.
+ */
+public class EventCommand extends Command {
+    private String requestContent;
```
Comment (szejiancheng) [inline src/main/java/duke/commands/EventCommand.java:10]: Seems like you prefer a line space between variables and methods, consider putting one here?

File: src/main/java/duke/parsers/Parser.java [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+     * @return A corresponding command.
+     */
+    public static Command parse(String fullCommand) throws DukeExceptions {
+        String[] splitStr = fullCommand.split(" ");
+        String requestContent = fullCommand.split(" ", 2).length == 2
+                ? fullCommand.split(" ", 2)[1]
+                : "";
+
+        //END
+        if (splitStr[0].equals("bye")) {
```
Comment (szejiancheng) [inline src/main/java/duke/parsers/Parser.java:26]: perhaps consider putting this as a "bye" case in your switch block?

File: src/main/java/duke/storage/Storage.java [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+
+            this.file = new File(path, "duke.txt");
+
+            if (!this.file.exists()) {
+                this.file.createNewFile();
+            }
+
+
+        } catch (IOException exception) {
+            System.out.println(exception.getMessage());
```
Comment (szejiancheng) [inline src/main/java/duke/storage/Storage.java:49]: So what happens when an exception is encountered during the try block? Where is it thrown to, and can the program be recovered?

File: src/main/java/duke/tasklist/TaskList.java [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+    /**
+     * Prints out all the task in the task list.
+     */
+    public void list() {
+        String result = "";
+        for (int i = 0; i < this.len; i++) {
+            int index = i + 1;
+            result += "  " + index + ". " + list[i].toString();
+        }
+        System.out.print(result);
```
Comment (szejiancheng) [inline src/main/java/duke/tasklist/TaskList.java:47]: Is there a way to allow the Ui class to handle all user input and output instead of printing from this class?

File: src/main/java/duke/tasks/Task.java [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+     * formats the task into a string to be saved locally
+     */
+    public abstract String saveTask();
+
+    /**
+     * decodes a task that has been formatted to be saved locally back into a task.
+     *
+     * @return the task that has been decoded
+     */
+    public static Task decode(String taskString) throws DukeExceptions {
```
Comment (szejiancheng) [inline src/main/java/duke/tasks/Task.java:64]: I've noticed that you have a parser class, how would you make it so that it handles more parsing rather than the Task class handling parsing?

**Review by szejiancheng — COMMENTED** [review]
Cool structure of Duke! I would suggest that the methods be tweaked such that each class is wholly responsible for one aspect of Duke's functionality for debugging and encapsulation purposes.

File: src/main/java/duke/commands/Command.java [hunk 13]
```diff
@@ -0,0 +1,40 @@
+package duke.commands;
```
Comment (benjamin-wee) [inline src/main/java/duke/commands/Command.java:1]: I like how you used sub packages to maintain your code!

File: src/main/java/duke/parsers/Parser.java (hunk 1 above)
Comment (benjamin-wee) [inline src/main/java/duke/parsers/Parser.java:61]: I like how you used a switch block as it makes you code much more readable and understandable! The form of the switch block also follows the coding standard!

File: src/main/java/duke/tasklist/TaskList.java [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+        Task deletedTask = list[index];
+        for (int i = index; i < this.len; i++) {
+            this.list[i] = this.list[i + 1];
+        }
+
+        this.len--;
+        String reply = "  Noted. I've removed this task:\n"
+                + "    " + deletedTask.toString()
+                + "  Now you have " + this.len + " tasks in the list.\n";
+        System.out.print(reply);
```
Comment (benjamin-wee) [inline src/main/java/duke/tasklist/TaskList.java]: Is it possible to allow the Ui class to handle user input and output here?

File: src/main/java/duke/tasks/Deadline.java [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+            throw new DukeExceptions("Format of date was not recognized, use YYYY-MM-DD and HH:MM");
+        }
+    }
+
+    /**
+     * Returns the deadline in a formatted string to be saved locally.
+     *
+     * @return the deadline in string format
+     */
+    public String saveTask() {
```
Comment (benjamin-wee) [inline src/main/java/duke/tasks/Deadline.java:42]: I like how the javadoc is descriptive and helps readers understand the code better!

**Review by benjamin-wee — COMMENTED** [review]
Good job in general making your code understandable and highly maintainable! Keep up the good work!
